libV2/client/index.js

```diff
@@ -0,0 +1,248 @@
+
+class Chunker extends Transform {
+    constructor(options) {
+        super({ objectMode: true, ...options });
+        this._chunkSize = options.chunksize || 100;
+        this._currentChunk = [];
+    }
+
+    _transform(chunk, encoding, callback) {
+        this._currentChunk.push(chunk);
+        if (this._currentChunk.length >= this._chunkSize) {
+            this.push(this._currentChunk);
+            this._currentChunk = [];
+        }
+        callback();
+    }
+
+    _flush(callback) {
+        if (this._currentChunk.length) {
+            this.push(this._currentChunk);
+        }
+        callback();
+    }
+}
+
+class Uploader extends Transform {
+    constructor(options) {
+        super({ objectMode: true, ...options });
+        this._ingest = options.ingest;
+    }
+
+    _transform(chunk, encoding, callback) {
+        this._ingest(chunk.map(i => new UtapiMetric(i.value)))
+            .then(() => {
+                this.push({
+                    success: true,
+                    keys: chunk.map(i => i.key),
+                });
+                callback();
+            },
+            error => {
+                this.push({
+                    success: false,
+                    keys: [],
+                });
+                moduleLogger.error('error uploading metrics from retry cache', { error });
+                callback();
+            });
+    }
+}
+
+class UtapiClient {
+    constructor(config) {
+        this._host = (config && config.host) || 'localhost';
+        this._port = (config && config.port) || '8100';
+        this._logger = (config && config.port) || moduleLogger;
+        this._maxCachedMetrics = (config && config.maxCachedMetrics) || 200000; // roughly 100MB
+        this._numCachedMetrics = 0;
+        this._retryCache = levelup(encode(memdown(), { valueEncoding: 'json' }));
+        this._drainTimer = null;
+        this._drainCanSchedule = true;
+        this._drainDelay = (config && config.drainDelay) || 30000;
+    }
+
+    async join() {
+        await this._flushRetryCacheToLogs();
+        this._retryCache.close();
+    }
+
+    async _pushToUtapi(metrics) {
+        const resp = await needle(
+            'post',
+            `http://${this._host}:${this._port}/v2/ingest`,
+            metrics.map(metric => metric.getValue()),
+            { json: true },
+        );
+        if (resp.statusCode !== 200) {
+            throw Error('failed to push metric, server returned non 200 status code',
+                { respCode: resp.statusCode, respMessage: resp.statusMessage });
+        }
+    }
+
+    async _addToRetryCache(metric) {
+        if (this._numCachedMetrics < this._maxCachedMetrics) {
+            try {
+                await this._retryCache.put(metric.uuid, metric.getValue());
+                this._numCachedMetrics += 1;
+                await this._scheduleDrain();
+                return true;
+            } catch (error) {
+                this._logger
```

Comment:
might be useful to log the `metric` object as well if not too big, although it's logged in the recovery log, it could maybe help to correlate this error with the recovery log.

Reply:
we can log the uuid of the metric and avoid logging the whole thing